envoy/upstream/eds_resources_cache.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include "envoy/common/optref.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/core/v3/config_source.pb.h"
+#include "envoy/config/endpoint/v3/endpoint.pb.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Upstream {
+
+// An interface for cached resource removed callback.
+class EdsResourceRemovalCallback {
+public:
+  virtual ~EdsResourceRemovalCallback() = default;
+
+  // Invoked when a cached resource is removed from the cache.
+  virtual void onCachedResourceRemoved(absl::string_view resource_name) PURE;
+};
+
+// This is an xDS resources cache for EDS subscriptions. It is a multi-level cache,
+// where the first level is a per-ConfigSource cache, and the second level is
+// per resource-name cache. In addition to saving resources in the cache,
+// retrieving them, and removing them, the cache allows attaching watchers to the
+// retrieved resources using ResourceRemovalCallback. The watchers will be notified when
+// the resource is removed from the cache.
+// This cache is needed to allow EDS subscriptions to use previously fetched
+// EDS resources, but only if no new EDS resource arrives. The removal callback
+// is used to ensure resource TTL is working properly.
+// TODO(adisuissa): In the future this can be a per-xDS resource-type cache.
+class EdsResourcesCache {
+public:
+  virtual ~EdsResourcesCache() = default;
+
+  class ConfigSourceResourcesMap {
+  public:
+    virtual ~ConfigSourceResourcesMap() = default;
+
+    /**
+     * Adds or updates a given resource name with its resource.
+     * Any watcher that was previously assigned to the resource will be removed
+     * without any notification.
```

Comment:
This might be a surprising use - why is `set` not triggering a `remove` callback if the resource changed? Maybe explain more the intent behind.